File: practicas/practica3_rb/Weapon.rb
#encoding: utf-8

#Clase que representa las armas de las que puede disponer una estación espacial para potenciar su energía al disparar

require_relative 'lib/WeaponToUI'

module Deepspace

  class Weapon

    #Constructor sin parámetros de la clase
    def initialize(name, type, uses)
      @name = name
      @type = type
      @uses = uses
    end

    #Constructor de copia de la clase
    def self.newCopy(orig)
      Weapon.new(orig.name, orig.type, orig.uses)
    end

    #Método consultor de nombre del arma
    def name
      @name
    end

    #Método consultor del tipo de arma
    def type
      @type
    end

    #Método consultor del número de usos del arma
    def uses
      @uses
    end

    #Método que devuelve la potencia de disparo del arma
    def power
     @type.power
    end


    #Método que decrementa uses en una unidad y devuelve el valor de power si uses es mayor que 0,
    #o que devuelve 1.0 en otro caso
    def useIt
      if @uses > 0
        @uses -= 1
        return power
      else
        return 1.0
      end
    end

    def to_s
      return "Name: " + @name.to_s + "\nType: " + @type.to_s + "\nUses: " + @uses.to_s
    end

    def getUIversion
      WeaponToUI.new self
    end

  end

end

File: practicas/practica5/src/main/java/deepspace/Weapon.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

/**
 * representa a las armas de las que puede disponer una estación espacial para potenciar suenergía al disparar.
 */
public class Weapon implements CombatElement, Copyable<Weapon>{
    private String name;
    private WeaponType type;
    private int uses;
    
    /**
     * constructor de la clase
     * @param name nombre del arma
     * @param type tipo del arma
     * @see WeaponType
     * @param uses cantidad de usos del arma
     */
    Weapon(String name, WeaponType type, int uses) {
        this.name = name;
        this.type = type;
        this.uses = uses;
    }
    
    /**
     * Constructor copia
     * @param w instancia a copiar
     */
    Weapon(Weapon w) {
        this (w.name, w.getType(), w.getUses());
    }
    
    /**
     * @return el tipo del arma
     * @see WeaponType
     */
    public WeaponType getType(){
        return type;
    }
    
    /**
     * @return usos restantes del arma
     */
    @Override
    public int getUses() {
        return uses;
    }
    
    /**
     * @return potencia del arma
     */
    public float power() {
        return type.getPower();
    }
    
    /**
     * Usa el arma disminuyendo sus usos en uno
     * @return el poder del arma o 1.0 en caso de que no queden usos
     */
    @Override
    public float useIt(){
        if(uses > 0){
            uses--;
            return power();
        }
        
        return 1.0f;
    }
    
    WeaponToUI getUIversion() {
        return new WeaponToUI(this);
    }

    @Override
    public String toString() {
        return "name: " + name + "\ntype: " + type.toString() + "\nuses: " + uses;
    }

    @Override
    public Weapon copy() {
        return new Weapon(this);
    }
    
    
}
